core/arch/aarch64/emit_utils.c

```diff
@@ -50,6 +50,69 @@
 /*                               EXIT STUB                                 */
 /***************************************************************************/
 
+/* We use multiple approaches to linking based on how far away the target
+ * fragment is:
+ *
+ *     Unlinked:
+ *         exit_cti stub
+ *         ...
+ *       stub:
+ *         stp  x0, x1, [x28]
+ *         movz x0, #&linkstub[0, 16),  lsl #0x00
+ *         movk x0, #&linkstub[16, 32), lsl #0x10
+ *         movk x0, #&linkstub[32, 48), lsl #0x20
+ *         movk x0, #&linkstub[48, 64), lsl #0x30
+ *         ldr  x1, [#8/#12]
+ *         br   x1
+ *         <fcache-return>
+ *
+ *     Linked, exit_cti_reaches_target (near fragment):
+ *         exit_cti target_fragment
+ *         ...
+ *       stub:
+ *         stp  x0, x1, [x28]
+ *         movz x0, #&linkstub[0, 16),  lsl #0x00
+ *         movk x0, #&linkstub[16, 32), lsl #0x10
+ *         movk x0, #&linkstub[32, 48), lsl #0x20
+ *         movk x0, #&linkstub[48, 64), lsl #0x30
+ *         ldr  x1, [#8/#12]
+ *         br   x1
+ *         <fcache-return>
+ *
+ *     Linked, unconditional branch reaches target (intermediate fragment):
+ *         exit_cti stub
+ *         ...
+ *       stub:
+ *         b    target_fragment
+ *         movz x0, #&linkstub[0, 16),  lsl #0x00
+ *         movk x0, #&linkstub[16, 32), lsl #0x10
+ *         movk x0, #&linkstub[32, 48), lsl #0x20
+ *         movk x0, #&linkstub[48, 64), lsl #0x30
+ *         ldr  x1, [#8/#12]
+ *         br   x1
+ *         <fcache-return>
+ *
+ *     Linked, !unconditional branch reaches target (far fragment):
+ *         exit_cti stub
+ *         ...
+ *       stub:
+ *         stp  x0, x1, [x28]
+ *         movz x0, #&linkstub[0, 16),  lsl #0x00
+ *         movk x0, #&linkstub[16, 32), lsl #0x10
+ *         movk x0, #&linkstub[32, 48), lsl #0x20
+ *         movk x0, #&linkstub[48, 64), lsl #0x30
+ *         ldr  x1, [#8/#12]
+ *         br   x1
+ *         <target_fragment_prefix>
+ *
+ * To ensure atomicity of <target> patching, the data slot must be 8-byte
+ * aligned. We do this by reserving 12 bytes for the data slot and using the
+ * appropriate offset in ldr for the 8-byte aligned 8 byte region within it.
+ *
+ * For complete design details, see the following wiki
+ * https://github.com/DynamoRIO/dynamorio/wiki/Linking-Far-Fragments-on-AArch64
+ */
```

Comment:
Nice, I think this improves readability for future code maintainers (including future versions of ourselves).

Reply:
Definitely. Thanks!